Sources/NIOHPACK/HPACKEncoder.swift

```diff
@@ -0,0 +1,299 @@
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// An `HPACKEncoder` maintains its own dynamic header table and uses that to
+/// encode HTTP headers to an internal byte buffer.
+///
+/// This encoder functions as an accumulator: each encode operation will append
+/// bytes to a buffer maintained by the encoder, which must be cleared using
+/// `reset()` before the encode can be re-used. It maintains a header table for
+/// outbound header indexing, and will update the header table as described in
+/// RFC 7541, appending and evicting items as described there.
+public struct HPACKEncoder {
+    /// The default size of the encoder's dynamic header table.
+    public static var defaultDynamicTableSize: Int { return DynamicHeaderTable.defaultSize }
+    private static let defaultDataBufferSize = 128
+    
+    public struct HeaderDefinition {
+        var name: String
+        var value: String
+        var indexing: HPACKIndexing
+    }
+    
+    private enum EncoderState {
+        case idle(maxTableSize: Int)
+        case resized(currentMaxTableSize: Int, smallestMaxTableSize: Int?)
+        case encoding(maxTableSize: Int)
+    }
+    
+    // private but tests
+    var headerIndexTable: IndexedHeaderTable
+    
+    private var state: EncoderState
+    private var buffer: ByteBuffer!
+    
+    /// Whether to use Huffman encoding.
+    public let useHuffmanEncoding: Bool
+    
+    /// The current size of the dynamic table.
+    ///
+    /// This is defined as the sum of [name] + [value] + 32 for each header.
+    public var dynamicTableSize: Int {
+        return self.headerIndexTable.dynamicTableLength
+    }
+    
+    /// The current maximum size to which the dynamic header table may grow.
+    public private(set) var allowedDynamicTableSize: Int {
+        get { return self.headerIndexTable.dynamicTableAllowedLength }
+        set { self.headerIndexTable.dynamicTableAllowedLength = newValue }
+    }
+    
+    /// The hard maximum size of the dynamic header table, set via an HTTP/2
+    /// SETTINGS frame.
+    public var maximumDynamicTableSize: Int {
+        get { return self.headerIndexTable.maxDynamicTableLength }
+        set { self.headerIndexTable.maxDynamicTableLength = newValue }
+    }
+    
+    /// Sets the maximum size for the dynamic table and encodes the new value
+    /// at the start of the current packed header block to send to the peer.
+    ///
+    /// - note: The size cannot grow beyond the maximum allotted by the most recent
+    ///         acknowledged SETTINGS_HEADER_TABLE_SIZE. It will truncate to that
+    ///         value automatically.
```

Comment:
Do we think truncation is the right thing to do here?

Reply:
I guess, since the method is already throwing, that it can throw an error in that case, too.